### PR TITLE
feat(provider): add support for Vultr

### PR DESCRIPTION
**Pull request by amroessam**
First time writing go, this was easy thanks to great docs and example 🙏🏽 
Go easy on the roast 😛 

[Vultr DNS API docs for ref](https://www.vultr.com/api/#tag/dns/operation/create-dns-domain-record)

Excited for the merge, so I can use this instead of my janky script https://github.com/amroessam/vultr-ddns

### CODE REVIEW

**Comment by amroessam** [comment]
Preemptively answering [these](https://github.com/qdm12/ddns-updater/pull/810#pullrequestreview-2292060445) questions

1. Yes, the default behavior is to create the record if it doesn't exist
2. Yes, I've tested it
3. Wildcards are supported but only for 1 record type, if it errors out it should reflect the error

**Review by ehab-qadah-lean — APPROVED** [review]

**Comment by amroessam** [comment]
@qdm12 any luck getting to this? 🙏🏽

File: .gitignore [hunk 1]
```diff
@@ -1 +1,4 @@
-/data
\ No newline at end of file
+/data
+.vscode
+app
+cmd/ddns-updater/data/
```
Comment (qdm12) [inline .gitignore:4]: Revert this 🙏  (or do another PR if you want to add i.e. `.vscode` to the .gitignore)

File: docs/vultr.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# Variomedia
```
Comment (qdm12) [inline docs/vultr.md:1]: The typical mistake I do 50% of the time 😄 

```suggestion
# Vultr
```

File: docs/vultr.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+### Example
+
+```json
+{
+  "settings": [
+    {
+      "provider": "vultr",
+      "domain": "example.com",
+      "owner": "potato",
```
Comment (qdm12) [inline docs/vultr.md:13]: The last program works with `domain` only, the owner (here `potato`) is automatically extracted from the domain string now.

```suggestion
      "domain": "potato.example.com",
```

File: docs/vultr.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+### Compulsory parameters
+
+- `"domain"` is the domain to update. It can be `example.com` (root domain) or `potato.example.com` (subdomain of `example.com`).
+- `"token"` is your API key, can be obtained from the [account settings](https://my.vultr.com/settings/#settingsapi), this is used as a bearer token to authenticate your request
+
+### Optional parameters
+
+- `"ip_version"` can be `ipv4` (A records), or `ipv6` (AAAA records) or `ipv4 or ipv6` (update one of the two, depending on the public ip found). It defaults to `ipv4 or ipv6`.
+- `"ipv6_suffix"` is the IPv6 interface identifier suffix to use. It can be for example `0:0:0:0:72ad:8fbb:a54e:bedd/64`. If left empty, it defaults to no suffix and the raw public IPv6 address obtained is used in the record updating.
+- `"ttl"` default is 900
```
Comment (qdm12) [inline docs/vultr.md:31]: Are you sure it's 900?

File: docs/vultr.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      "ttl": 300,
+      "ip_version": "ipv4"
+    }
+  ]
+}
+```
+
+### Compulsory parameters
+
+- `"domain"` is the domain to update. It can be `example.com` (root domain) or `potato.example.com` (subdomain of `example.com`).
```
Comment (qdm12) [inline docs/vultr.md:24]: Should the wildcard be documented as well, indicating the limitation on record type?

File: internal/provider/providers/vultr/provider.go [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+	ipVersion  ipversion.IPVersion
+	ipv6Suffix netip.Prefix
+	token      string
+	ttl        uint32
+}
+
+func New(data json.RawMessage, domain, owner string,
+	ipVersion ipversion.IPVersion, ipv6Suffix netip.Prefix) (
+	provider *Provider, err error) {
+
```
Comment (qdm12) [inline internal/provider/providers/vultr/provider.go:32]: nit no need for first empty line

```suggestion
```

File: internal/provider/providers/vultr/provider.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+func validateSettings(domain, owner, token string, ttl uint32) (err error) {
+	err = utils.CheckDomain(domain)
+	if err != nil {
+		return fmt.Errorf("%w: %w", errors.ErrDomainNotValid, err)
+	}
+
+	switch {
+	case owner == "":
+		return fmt.Errorf("%w", errors.ErrOwnerNotSet)
```
Comment (qdm12) [inline internal/provider/providers/vultr/provider.go:65]: nit I believe this can be removed, since it defaults to `@` if it's not set (in code called before creating this provider). Feel free to double check though.


```suggestion
```

File: internal/provider/providers/vultr/provider.go [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+	case token == "":
+		return fmt.Errorf("%w", errors.ErrAPIKeyNotSet)
```
Comment (qdm12) [inline internal/provider/providers/vultr/provider.go:67]: Isn't there an `ErrTokenNotSet`? If not, please create one 🙏 Otherwise it's rather confusing for the user, since it's a `token` and the error talks about an api key.

File: internal/provider/providers/vultr/provider.go [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+	case ttl == 0:
+		return fmt.Errorf("%w", errors.ErrTTLNotSet)
```
Comment (qdm12) [inline internal/provider/providers/vultr/provider.go:69]: ttl is not required in API requests, so allow it to be zero, and just don't send it with the request if it's zero.

File: internal/provider/providers/vultr/provider.go [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+		Scheme: "https",
+		Host:   "api.vultr.com",
+		Path:   fmt.Sprintf("/v2/domains/%s/records", p.domain),
+	}
+
+	requestData := struct {
+		Type string `json:"type"`
+		IP   string `json:"data"`
+		Name string `json:"name"`
+		TTL  uint32 `json:"ttl"`
```
Comment (qdm12) [inline internal/provider/providers/vultr/provider.go:134]: So that the ttl field is not set if its value is zero.

```suggestion
		TTL  uint32 `json:"ttl,omitempty"`
```

File: internal/provider/providers/vultr/provider.go [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+	err = decoder.Decode(&parsedJSON)
+	if err != nil {
+		return netip.Addr{}, fmt.Errorf("json decoding response body: %w", err)
+	}
+
+	if parsedJSON.Error != "" {
+		return netip.Addr{}, fmt.Errorf("API Error: %s", parsedJSON.Error)
+	}
+
+	if response.StatusCode != http.StatusCreated {
```
Comment (qdm12) [inline internal/provider/providers/vultr/provider.go:182]: Oh it's always statusCreated, never StatusOk (200)?

File: internal/provider/providers/vultr/provider.go [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+		return netip.Addr{}, fmt.Errorf("%w: %d: %s",
+			errors.ErrHTTPStatusNotValid, response.StatusCode, utils.BodyToSingleLine(response.Body))
+	}
```
Comment (qdm12) [inline internal/provider/providers/vultr/provider.go:185]: I feel like this code would never trigger because we check the parsed error before it. Maybe have something like this instead:

```go
	if response.StatusCode != http.StatusCreated {
		return netip.Addr{}, fmt.Errorf("%w: %d: %s",
			errors.ErrHTTPStatusNotValid, response.StatusCode, parsedJSON.Error)
	}

	if parsedJSON.Error != "" {
		return netip.Addr{}, fmt.Errorf("%w: %s", errors.ErrUnsuccessful, parsedJSON.Error)
	}
```

Also for your own knowledge, note the previously used `utils.BodyToSingleLine(response.Body)` would not work because it has already been read by the JSON decoder. One way to work around this is to io.ReadAll(response.Body) to have it in a byte slice, then use `json.Unmarshal(byteSlice, &parsedJSON)` and you can then re-use `byteSlice` after for example with `utils.ToSingleLine(string(byteSlice))`. Not something to do here really I think though 😄

**Review by qdm12 — CHANGES_REQUESTED** [review]
Hi there! Thanks for the PR and congrats on your first few Go steps! 💯 
Looking pretty good, with a few minor changes to make 😉

File: docs/vultr.md (hunk 2 above)
Comment (amroessam) [reply docs/vultr.md:1]: Fixing

File: docs/vultr.md (hunk 4 above)
Comment (amroessam) [reply docs/vultr.md:30]: Yes, I've tested it

File: docs/vultr.md (hunk 5 above)
Comment (amroessam) [reply docs/vultr.md:23]: Good shout, let me do that

File: internal/provider/providers/vultr/provider.go (hunk 11 above)
Comment (amroessam) [reply internal/provider/providers/vultr/provider.go:179]: Yes sir ✅

File: internal/provider/providers/vultr/provider.go (hunk 8 above)
Comment (amroessam) [reply internal/provider/providers/vultr/provider.go:64]: This is a bit confusing, the token is actually a bearer token, but from their settings page they're asking for an API key, I think APIKeyNotSet makes more sense as that's what the user is asked to send.

File: internal/provider/providers/vultr/provider.go (hunk 12 above)
Comment (amroessam) [reply internal/provider/providers/vultr/provider.go:180]: I can't flip it as that would return, their api returns an error string if it's not 201, so that checks for the error message and displays it for the user. If the object doesn't contain an error though, we can rely on the status code. 

Thanks for your tip

File: .gitignore (hunk 1 above)
Comment (amroessam) [reply .gitignore]: #835 handled here

File: internal/provider/providers/vultr/provider.go (hunk 8 above)
Comment (amroessam) [reply internal/provider/providers/vultr/provider.go]: I've updated to be token not set, let me know what you prefer though 🙏🏽

Comment (amroessam) [reply internal/provider/providers/vultr/provider.go]: image

Comment (qdm12) [reply internal/provider/providers/vultr/provider.go]: Oh ok, let's match what vultr has, so instead rename that token field to `apikey` and use that previous apikey not set error. The user shouldn't (and probably doesn't) care about the implementation details such as how it's sent to the API (bearer token).

File: internal/provider/providers/vultr/provider.go (hunk 12 above)
Comment (amroessam) [reply internal/provider/providers/vultr/provider.go:180]: There are 3 cases
1. 201 created
2. HTTP error (400, 401, 403, 404) with error message
3. No documentation about 5xx where probably there won't be an error message

So I first check the error message if it exists, and if not, I check the status code. Does that make sense?

File: internal/provider/providers/vultr/provider.go (hunk 8 above)
Comment (amroessam) [reply internal/provider/providers/vultr/provider.go:64]: Fixed, take a look pls 🙏🏽

File: internal/provider/providers/vultr/provider.go (hunk 12 above)
Comment (qdm12) [reply internal/provider/providers/vultr/provider.go:180]: Ah interesting, so the code needs more change then.

I suggest you have a little function:

```go
func parseJSONError(body io.Reader) (message string) {
	var parsedJSON struct {
		Error string `json:"error"`
	}
	decoder := json.NewDecoder(body)
	err := decoder.Decode(&parsedJSON)
	if err != nil {
		return "<json decoding error: " + err.Error() + ">"
	}
	return parsedJSON.Error
}
```

and then have a switch on the status code:


```go
	defer response.Body.Close()
	
	switch response.StatusCode {
	case http.StatusCreated:
	case http.StatusBadRequest:
		return fmt.Errorf("%w: %s", errors.ErrBadRequest, parseJSONError(response.Body))
	case http.StatusUnauthorized, http.StatusForbidden:
		return fmt.Errorf("%w: %s", errors.ErrAuth, parseJSONError(response.Body))
	case http.StatusNotFound:
		return fmt.Errorf("%w: %s", errors.ErrDomainNotFound, parseJSONError(response.Body))
	default:
		return netip.Addr{}, fmt.Errorf("%w: %s: %s",	errors.ErrHTTPStatusNotValid,
			response.Status, utils.BodyToSingleLine(response.Body))
	}

	decoder := json.NewDecoder(response.Body)
	var parsedJSON struct {
		Error string
		Status uint32
		Record struct {
			ID   string `json:"id"`
			Type string `json:"type"`
			IP   string `json:"data"`
			Name string `json:"name"`
			TTL  uint32 `json:"ttl"`
		}
	}
	err = decoder.Decode(&parsedJSON)
	if err != nil {
		return netip.Addr{}, fmt.Errorf("json decoding response body: %w", err)
	}
```

File: internal/provider/providers/vultr/provider.go [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+	response, err := client.Do(request)
+	if err != nil {
+		return netip.Addr{}, err
+	}
+	defer response.Body.Close()
+
+	decoder := json.NewDecoder(response.Body)
+	var parsedJSON struct {
+		Error string
+		Status uint32
```
Comment (qdm12) [inline internal/provider/providers/vultr/provider.go:159]: Also what is that Status used for? 🤔

File: internal/provider/providers/vultr/provider.go [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+	ipv6Suffix netip.Prefix
+	apiKey      string
+	ttl        uint32
+}
+
+func New(data json.RawMessage, domain, owner string,
+	ipVersion ipversion.IPVersion, ipv6Suffix netip.Prefix) (
+	provider *Provider, err error) {
+	var providerSpecificSettings struct {
+		APIKey string `json:"apiKey"`
```
Comment (qdm12) [inline internal/provider/providers/vultr/provider.go:33]:
```suggestion
		APIKey string `json:"apikey"`
```

File: docs/vultr.md [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+### Example
+
+```json
+{
+  "settings": [
+    {
+      "provider": "vultr",
+      "domain": "potato.example.com",
+      "apiKey": "AAAAAAAAAAAAAAA",
```
Comment (qdm12) [inline docs/vultr.md:13]: nit to match all-lowercase json elsewhere in this codebase

```suggestion
      "apikey": "AAAAAAAAAAAAAAA",
```

File: docs/vultr.md [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+      "ip_version": "ipv4"
+    }
+  ]
+}
+```
+
+### Compulsory parameters
+
+- `"domain"` is the domain to update. It can be `example.com` (root domain) or `potato.example.com` (subdomain of `example.com`). In case of a wildcard, this only works if there are no existing wildcard records of any type. If there is, it should error out.
+- `"apiKey"` is your API key, can be obtained from the [account settings](https://my.vultr.com/settings/#settingsapi), this is used as a bearer token to authenticate your request
```
Comment (qdm12) [inline docs/vultr.md:24]:
```suggestion
- `"apikey"` is your API key, can be obtained from the [account settings](https://my.vultr.com/settings/#settingsapi), this is used as a bearer token to authenticate your request
```

File: internal/provider/providers/vultr/provider.go (hunk 8 above)
Comment (qdm12) [reply internal/provider/providers/vultr/provider.go:64]: Thanks! I also love how they say "Personal Access Token" and also "API key"... I guess it doesn't matter too much 😄

File: internal/provider/providers/vultr/provider.go (hunk 13 above)
Comment (amroessam) [reply internal/provider/providers/vultr/provider.go:159]: It's the HTTP status code

File: internal/provider/providers/vultr/provider.go (hunk 14 above)
Comment (amroessam) [reply internal/provider/providers/vultr/provider.go:33]: Sure, but why not camelCase? 🤔

File: internal/provider/providers/vultr/provider.go (hunk 13 above)
Comment (qdm12) [reply internal/provider/providers/vultr/provider.go:159]: Ok that's weird 😄 Should we... remove it from that struct?? Since it's... well, in the HTTP response already...

**Comment by amroessam** [comment]
Sorry for the delay, I was away. Can you recheck?

File: internal/provider/providers/vultr/provider.go (hunk 13 above)
Comment (amroessam) [reply internal/provider/providers/vultr/provider.go]: For the sake of implementing the API, I've added it, can remove it if you insist 🙏🏽

File: internal/provider/providers/vultr/provider.go (hunk 14 above)
Comment (qdm12) [reply internal/provider/providers/vultr/provider.go]: Just to match other existing fields, which are all snake case. Also a bit of a habit to have snake case json keys.

**Comment by qdm12** [comment]
Hey there, thanks for the last few commits!
In the end I decided to push directly on your branch, since there were quite a few things to change (ranging from not important at all to somewhat important - see commits with `fix: ` prefix). Feel free to browse commit by commit, since I tried my best to separate changes one by one for knowledge sharing purposes!
I also rebased your branch on the master branch here, so feel free to get changes using `git checkout master; git fetch; git reset --hard origin/master` 😉 
Just waiting for the CI to pass and I'll merge this. Thanks!
